<a href="https://colab.research.google.com/github/evitts1989/DS-Unit-1-Sprint-2-Statistics/blob/master/Corey_Evitts_DS_Unit_1_Sprint_Challenge_2_(2).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Science Unit 1 Sprint Challenge 2

## Exploring Data, Testing Hypotheses

In this sprint challenge you will look at a dataset of people being approved or rejected for credit.

https://archive.ics.uci.edu/ml/datasets/Credit+Approval

Data Set Information: This file concerns credit card applications. All attribute names and values have been changed to meaningless symbols to protect confidentiality of the data. This dataset is interesting because there is a good mix of attributes -- continuous, nominal with small numbers of values, and nominal with larger numbers of values. There are also a few missing values.

Attribute Information:
- A1: b, a.
- A2: continuous.
- A3: continuous.
- A4: u, y, l, t.
- A5: g, p, gg.
- A6: c, d, cc, i, j, k, m, r, q, w, x, e, aa, ff.
- A7: v, h, bb, j, n, z, dd, ff, o.
- A8: continuous.
- A9: t, f.
- A10: t, f.
- A11: continuous.
- A12: t, f.
- A13: g, p, s.
- A14: continuous.
- A15: continuous.
- A16: +,- (class attribute)

Yes, most of that doesn't mean anything. A16 is a variable that indicates whether or not a person's request for credit has been approved or denied. This is a good candidate for a y variable since we might want to use the other features to predict this one. The remaining variables have been obfuscated for privacy - a challenge you may have to deal with in your data science career.

Sprint challenges are evaluated based on satisfactory completion of each part. It is suggested you work through it in order, getting each aspect reasonably working, before trying to deeply explore, iterate, or refine any given step. Once you get to the end, if you want to go back and improve things, go for it!

## Part 1 - Load and validate the data

- Load the data as a `pandas` data frame.
- Validate that it has the appropriate number of observations (you can check the raw file, and also read the dataset description from UCI).
- UCI says there should be missing data - check, and if necessary change the data so pandas recognizes it as na
- Make sure that the loaded features are of the types described above (continuous values should be treated as float), and correct as necessary

This is review, but skills that you'll use at the start of any data exploration. Further, you may have to do some investigation to figure out which file to load from - that is part of the puzzle.

In [0]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns
from scipy import stats

In [55]:
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data', header=None)
df.head()

0      1      2  3  4  5  6     7  8  9   10 11 12     13   14 15
0  b  30.83  0.000  u  g  w  v  1.25  t  t   1  f  g  00202    0  +
1  a  58.67  4.460  u  g  q  h  3.04  t  t   6  f  g  00043  560  +
2  a  24.50  0.500  u  g  q  h  1.50  t  f   0  f  g  00280  824  +
3  b  27.83  1.540  u  g  w  v  3.75  t  t   5  t  g  00100    3  +
4  b  20.17  5.625  u  g  w  v  1.71  t  f   0  f  s  00120    0  +

In [36]:
df[0].value_counts()

b    468
a    210
?     12
Name: 0, dtype: int64

In [0]:
df.replace('?', np.nan, inplace = True)

In [38]:
df.isnull().sum()

0     12
1     12
2      0
3      6
4      6
5      9
6      9
7      0
8      0
9      0
10     0
11     0
12     0
13    13
14     0
15     0
dtype: int64

In [0]:
pos_df = df.loc[df[15] == '+']
neg_df = df.loc[df[15] == '-']

## Part 2 - Exploring data, Testing hypotheses

The only thing we really know about this data is that A16 is the class label. Besides that, we have 6 continuous (float) features and 9 categorical features.

Explore the data: you can use whatever approach (tables, utility functions, visualizations) to get an impression of the distributions and relationships of the variables. In general, your goal is to understand how the features are different when grouped by the two class labels (`+` and `-`).

For the 6 continuous features, how are they different when split between the two class labels? Choose two features to run t-tests (again split by class label) - specifically, select one feature that is *extremely* different between the classes, and another feature that is notably less different (though perhaps still "statistically significantly" different). You may have to explore more than two features to do this.

For the categorical features, explore by creating "cross tabs" (aka [contingency tables](https://en.wikipedia.org/wiki/Contingency_table)) between them and the class label, and apply the Chi-squared test to them. [pandas.crosstab](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.crosstab.html) can create contingency tables, and [scipy.stats.chi2_contingency](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html) can calculate the Chi-squared statistic for them.

There are 9 categorical features - as with the t-test, try to find one where the Chi-squared test returns an extreme result (rejecting the null that the data are independent), and one where it is less extreme.

**NOTE** - "less extreme" just means smaller test statistic/larger p-value. Even the least extreme differences may be strongly statistically significant.

Your *main* goal is the hypothesis tests, so don't spend too much time on the exploration/visualization piece. That is just a means to an end - use simple visualizations, such as boxplots or a scatter matrix (both built in to pandas), to get a feel for the overall distribution of the variables.

This is challenging, so manage your time and aim for a baseline of at least running two t-tests and two Chi-squared tests before polishing. And don't forget to answer the questions in part 3, even if your results in this part aren't what you want them to be.

In [40]:
df.shape

(690, 16)

In [41]:
neg_df.shape

(383, 16)

In [42]:
pos_df.shape

(307, 16)

In [43]:
pos_df.describe()

2           7           10             14
count  307.000000  307.000000  307.000000     307.000000
mean     5.904951    3.427899    4.605863    2038.859935
std      5.471485    4.120792    6.320242    7659.763941
min      0.000000    0.000000    0.000000       0.000000
25%      1.500000    0.750000    0.000000       0.000000
50%      4.460000    2.000000    3.000000     221.000000
75%      9.520000    5.000000    7.000000    1209.000000
max     28.000000   28.500000   67.000000  100000.000000

In [44]:
neg_df.describe()

2           7           10           14
count  383.000000  383.000000  383.000000   383.000000
mean     3.839948    1.257924    0.631854   198.605744
std      4.337662    2.120481    1.900049   671.608839
min      0.000000    0.000000    0.000000     0.000000
25%      0.835000    0.125000    0.000000     0.000000
50%      2.210000    0.415000    0.000000     1.000000
75%      5.000000    1.500000    0.000000    67.000000
max     26.335000   13.875000   20.000000  5552.000000

In [45]:
pos_df.head()

0      1      2  3  4  5  6     7  8  9   10 11 12     13   14 15
0  b  30.83  0.000  u  g  w  v  1.25  t  t   1  f  g  00202    0  +
1  a  58.67  4.460  u  g  q  h  3.04  t  t   6  f  g  00043  560  +
2  a  24.50  0.500  u  g  q  h  1.50  t  f   0  f  g  00280  824  +
3  b  27.83  1.540  u  g  w  v  3.75  t  t   5  t  g  00100    3  +
4  b  20.17  5.625  u  g  w  v  1.71  t  f   0  f  s  00120    0  +

In [46]:
neg_df.head()

0      1    2  3  4   5   6       7  8  9   10 11 12     13  14 15
70  b  32.33  7.5  u  g   e  bb   1.585  t  f   0  t  s  00420   0  -
71  b  34.83  4.0  u  g   d  bb  12.500  t  f   0  t  g    NaN   0  -
72  a  38.58  5.0  u  g  cc   v  13.500  t  f   0  t  g  00980   0  -
73  b  44.25  0.5  u  g   m   v  10.750  t  f   0  f  s  00400   0  -
74  b  44.83  7.0  y  p   c   v   1.625  f  f   0  f  g  00160   2  -

## Test 1: a11 (continuous)

In [49]:
test1_pos = pos_df[[10]]
test1_pos.head()

10
0   1
1   6
2   0
3   5
4   0

In [50]:
test1_neg = neg_df[[10]]
test1_neg.head()

10
70   0
71   0
72   0
73   0
74   0

In [51]:
test1_pos.describe()

10
count  307.000000
mean     4.605863
std      6.320242
min      0.000000
25%      0.000000
50%      3.000000
75%      7.000000
max     67.000000

In [52]:
test1_neg.describe()

10
count  383.000000
mean     0.631854
std      1.900049
min      0.000000
25%      0.000000
50%      0.000000
75%      0.000000
max     20.000000

In [0]:
from scipy import stats

**Test 1 Hypothesis**:
The Mean of Column A11 for people approved seems higher than the mean of people declined.
**𝐻0**  = Column A11 was not a factor used to determine if someone was approved or denied credit
**𝐻𝑎 ** = Column A11 was a factor that contributed to someone's approval for credit

In [56]:
print(stats.ttest_ind(test1_pos,test1_neg).statistic)

[11.66700422]


In [57]:
print(stats.ttest_ind(test1_pos,test1_neg).pvalue)

[7.95771857e-29]


### Mean for accepted: 4.6
### Standard Deviation: 6.32
### Mean for those declined: 0.63
### Standard Deviation: 1.90
## The T Score is 11.66!
## The P Value is [7.95771857e-29] - This is an EXTREMELY small number.
##The probability that a difference this large was due to pure chance is so low it has to be expressed in scientific notation.
# I reject the Null Hypothesis. A given person's A11 Value was a factor in deciding credit approval or disapproval.

##Test 2

In [58]:
pos_df.head()

0      1      2  3  4  5  6     7  8  9   10 11 12     13   14 15
0  b  30.83  0.000  u  g  w  v  1.25  t  t   1  f  g  00202    0  +
1  a  58.67  4.460  u  g  q  h  3.04  t  t   6  f  g  00043  560  +
2  a  24.50  0.500  u  g  q  h  1.50  t  f   0  f  g  00280  824  +
3  b  27.83  1.540  u  g  w  v  3.75  t  t   5  t  g  00100    3  +
4  b  20.17  5.625  u  g  w  v  1.71  t  f   0  f  s  00120    0  +

In [59]:
neg_df.head()

0      1    2  3  4   5   6       7  8  9   10 11 12     13  14 15
70  b  32.33  7.5  u  g   e  bb   1.585  t  f   0  t  s  00420   0  -
71  b  34.83  4.0  u  g   d  bb  12.500  t  f   0  t  g    NaN   0  -
72  a  38.58  5.0  u  g  cc   v  13.500  t  f   0  t  g  00980   0  -
73  b  44.25  0.5  u  g   m   v  10.750  t  f   0  f  s  00400   0  -
74  b  44.83  7.0  y  p   c   v   1.625  f  f   0  f  g  00160   2  -

In [60]:
test2_pos_trial = pos_df[[1]]
test2_pos_trial.head(50)

1
0   30.83
1   58.67
2   24.50
3   27.83
4   20.17
5   32.08
6   33.17
7   22.92
8   54.42
9   42.50
10  22.08
11  29.92
12  38.25
13  48.08
14  45.83
15  36.67
16  28.25
17  23.25
18  21.83
19  19.17
20  25.00
21  23.25
22  47.75
23  27.42
24  41.17
25  15.83
26  47.00
27  56.58
28  57.42
29  42.08
30  29.25
31  42.00
32  49.50
33  36.75
34  22.58
35  27.83
36  27.25
37  23.00
38  27.75
39  54.58
40  34.17
41  28.92
42  29.67
43  39.58
44  56.42
45  54.33
46  41.00
47  31.92
48  41.50
49  23.92

In [61]:
test2_pos_trial.isna().sum()

1    2
dtype: int64

In [0]:
test2_pos_trial = test2_pos_trial.dropna()

In [63]:
test2_pos_trial.isna().sum()

1    0
dtype: int64

In [64]:
test2_neg_trial = neg_df[[1]]
test2_neg_trial.head(10)

1
70  32.33
71  34.83
72  38.58
73  44.25
74  44.83
75  20.67
76  34.08
77  19.17
78  21.67
79  21.50

In [65]:
test2_neg_trial.isna().sum()

1    10
dtype: int64

In [0]:
test2_neg_trial = test2_neg_trial.dropna()

In [67]:
test2_neg_trial.isna().sum()

1    0
dtype: int64

In [68]:
test2_neg_trial.shape

(373, 1)

In [69]:
test2_neg_trial.describe()

1
count     373
unique    231
top     23.58
freq        6

In [0]:
test2_pos = test2_pos_trial
test2_neg = test2_neg_trial

In [72]:
test2_pos.head()

1
0  30.83
1  58.67
2  24.50
3  27.83
4  20.17

In [73]:
test2_neg.head()

1
70  32.33
71  34.83
72  38.58
73  44.25
74  44.83

In [74]:
test2_pos.shape

(305, 1)

In [75]:
test2_neg.shape

(373, 1)

In [0]:
test2_pos[[1]] = test2_pos[[1]].astype(float)
test2_neg[[1]] = test2_neg[[1]].astype(float)

# Test 2 Hypothesis
### The mean of column A2 appears to be almost the same between people accepted and people declined on the application.

##$H_0$ = Column A2 was not a factor used to determine if someone was approved or denied credit

##$H_a$ = Column A2 was a factor that contributed to someone's approval for credit



In [77]:
test2_neg.describe()

1
count  373.000000
mean    29.808231
std     10.919291
min     15.170000
25%     22.000000
50%     27.330000
75%     34.830000
max     80.250000

In [78]:
print(stats.ttest_ind(test2_pos,test2_neg).statistic)

[4.29221562]


In [79]:
print(stats.ttest_ind(test2_pos,test2_neg).pvalue)

[2.02766371e-05]


### Mean for those accepted: 33.7
### Standard Deviation: 12.8

### Mean for those declined: 29.8
### Standard Deviation: 10.91
## The T Score is 4.292!
## The P Value is [2.02766371e-05]- This is an EXTREMELY small number.
##The probability that a difference this large was due to pure chance is so low it has to be expressed in scientific notation.
# I reject the Null Hypothesis. A given person's A2 Value was a factor in deciding credit approval or disapproval.

#Test:3
##A12



In [80]:
pos_df.head()

0      1      2  3  4  5  6     7  8  9   10 11 12     13   14 15
0  b  30.83  0.000  u  g  w  v  1.25  t  t   1  f  g  00202    0  +
1  a  58.67  4.460  u  g  q  h  3.04  t  t   6  f  g  00043  560  +
2  a  24.50  0.500  u  g  q  h  1.50  t  f   0  f  g  00280  824  +
3  b  27.83  1.540  u  g  w  v  3.75  t  t   5  t  g  00100    3  +
4  b  20.17  5.625  u  g  w  v  1.71  t  f   0  f  s  00120    0  +

In [81]:
neg_df.head()

0      1    2  3  4   5   6       7  8  9   10 11 12     13  14 15
70  b  32.33  7.5  u  g   e  bb   1.585  t  f   0  t  s  00420   0  -
71  b  34.83  4.0  u  g   d  bb  12.500  t  f   0  t  g    NaN   0  -
72  a  38.58  5.0  u  g  cc   v  13.500  t  f   0  t  g  00980   0  -
73  b  44.25  0.5  u  g   m   v  10.750  t  f   0  f  s  00400   0  -
74  b  44.83  7.0  y  p   c   v   1.625  f  f   0  f  g  00160   2  -

In [0]:
test3_pos_trial = pos_df[[11]]
test3_neg_trial = neg_df[[11]]

In [83]:
test3_pos_trial.head()

11
0  f
1  f
2  f
3  t
4  f

In [86]:
test3_neg_trial.head()

11
70  t
71  t
72  t
73  f
74  f

In [87]:
test3_pos_trial.replace('f', 0, inplace = True)
test3_pos_trial.replace('t', 1, inplace = True)

/usr/local/lib/python3.6/dist-packages/pandas/core/frame.py:4278: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  method=method,


In [88]:
test3_pos_trial.head()

11
0   0
1   0
2   0
3   1
4   0

In [89]:
test3_neg_trial.replace('f', 0, inplace = True)
test3_neg_trial.replace('t', 1, inplace = True)

/usr/local/lib/python3.6/dist-packages/pandas/core/frame.py:4278: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  method=method,


In [90]:
test3_neg_trial.head()

11
70   1
71   1
72   1
73   0
74   0

In [0]:
test3_pos = test3_pos_trial
test3_neg = test3_neg_trial

In [92]:
test3_pos[[11]] = test3_pos[[11]].astype(float)
test3_neg[[11]] = test3_neg[[11]].astype(float)

/usr/local/lib/python3.6/dist-packages/pandas/core/frame.py:3509: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[k1] = value[k2]


In [93]:
test3_pos.head()

11
0  0.0
1  0.0
2  0.0
3  1.0
4  0.0

In [94]:
test3_neg.head()

11
70  1.0
71  1.0
72  1.0
73  0.0
74  0.0

In [95]:
test3_pos.describe()

11
count  307.000000
mean     0.475570
std      0.500218
min      0.000000
25%      0.000000
50%      0.000000
75%      1.000000
max      1.000000

In [96]:
test3_neg.describe()

11
count  383.000000
mean     0.443864
std      0.497489
min      0.000000
25%      0.000000
50%      0.000000
75%      1.000000
max      1.000000

In [97]:
test3_pos.rename(columns={11: 'A12'}, inplace=True)

/usr/local/lib/python3.6/dist-packages/pandas/core/frame.py:4238: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().rename(**kwargs)


In [98]:
test3_pos.head()

A12
0  0.0
1  0.0
2  0.0
3  1.0
4  0.0

In [99]:
test3_neg.rename(columns={11: 'A12'}, inplace=True)

/usr/local/lib/python3.6/dist-packages/pandas/core/frame.py:4238: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().rename(**kwargs)


In [100]:
test3_neg.head()

A12
70  1.0
71  1.0
72  1.0
73  0.0
74  0.0

# Test 3 Hypothesis
### The mean of column A12 appears to be relatively similar between people accepted and people declined on the application.
##$H_0$ = Column A12 was not a factor used to determine if someone was approved or denied credit
##$H_a$ = Column A12 was a factor that contributed to someone's approval for credit

In [0]:
# One-way chi square test
from scipy.stats import chisquare  

In [0]:
from scipy.stats import chisquare  # One-way chi square test

In [103]:
pd.crosstab(df[11],df[15])

15    +    -
11          
f   161  213
t   146  170

In [0]:
test3_xtab = pd.crosstab(df[11],df[15])

In [107]:
print(chisquare(test3_xtab, axis=None))

Power_divergenceResult(statistic=14.382608695652173, pvalue=0.0024280203352694887)


In [0]:
test3_xtab.replace('f', 0, inplace = True)
test3_xtab.replace('t', 1, inplace = True)

In [109]:
print(chisquare(test3_xtab, axis=None))

Power_divergenceResult(statistic=14.382608695652173, pvalue=0.0024280203352694887)


In [110]:
test3_xtab.head()

15    +    -
11          
f   161  213
t   146  170

#The p-value is very low - the likelyhood of seeing a result this extreme or more extreme is very low.
## I reject the null hypothesis, there is a chi square of 14

#Test 4 Pre-Processing
##A5


In [0]:
clean_df = df.dropna()

In [113]:
clean_df.head()

0      1      2  3  4  5  6     7  8  9   10 11 12     13   14 15
0  b  30.83  0.000  u  g  w  v  1.25  t  t   1  f  g  00202    0  +
1  a  58.67  4.460  u  g  q  h  3.04  t  t   6  f  g  00043  560  +
2  a  24.50  0.500  u  g  q  h  1.50  t  f   0  f  g  00280  824  +
3  b  27.83  1.540  u  g  w  v  3.75  t  t   5  t  g  00100    3  +
4  b  20.17  5.625  u  g  w  v  1.71  t  f   0  f  s  00120    0  +

In [0]:
test4_xtab = pd.crosstab(df[4],df[15])

In [115]:
test4_xtab.head()

15    +    -
4           
?     4    2
g   256  263
gg    2    0
p    45  118

# Test 4 Hypothesis
### column A5
##$H_0$ = Column A5 was not a factor used to determine if someone was approved or denied credit
##$H_a$ = Column A5 was a factor that contributed to someone's approval for credit

In [116]:
print(chisquare(test4_xtab, axis=None))

Power_divergenceResult(statistic=1056.9913043478261, pvalue=5.825554166604218e-224)


#There is a very high test statistic and a very small P value  - this means we can reject the null hypothesis, it's very unlikely this was due to chance.

## Part 3 - Analysis and Interpretation

Now that you've looked at the data, answer the following questions:

- Interpret and explain the two t-tests you ran - what do they tell you about the relationships between the continuous features you selected and the class labels?
- Interpret and explain the two Chi-squared tests you ran - what do they tell you about the relationships between the categorical features you selected and the class labels?
- What was the most challenging part of this sprint challenge?

Answer with text, but feel free to intersperse example code/results or refer to it from earlier.

###I have explained the relationships in part 2 along with the actual tests, oddly enough it seemed like every attribute on the credit application played a significant role in deciding whether a person was accepted or declined... I guess that makes sense. I'm sure within the data there exists an opposite case somewhere though for sure.

###This enstire sprint has been challenging because i missed some of the lecture and i dont fully grasp all of the concepts. The chi squared test seem to be where i have the most trouble.. 

# Stretch Goals: 
Do these to get a 3. These are not required in order to pass the Sprint Challenge.

## Part 1: 

Make sure that all of your dataframe columns have the appropriate data types. *Hint:* If a column has the datatype of "object" even though it's made up of float or integer values, you can coerce it to act as a numeric column by using the `pd.to_numeric()` function. In order to get a 3 on this section make sure that your data exploration is particularly well commented, easy to follow, and thorough.

## Part 2:

Write functions that can calculate t-tests and chi^2 tests on all of the appropriate column combinations from the dataset. (Remember that certain tests require certain variable types.)

## Part 3: 

Calculate and report confidence intervals on your most important mean estimates (choose at least two). Make some kind of a graphic or visualization to help us see visually how precise these estimates are.

In [0]:
# You can work the stretch goals down here or back up in their regular sections
# just make sure that they are labeled so that we can easily differentiate
# your main work from the stretch goals.